### PR TITLE
feat(gemini): autogenerated update

### PR DESCRIPTION
**Pull request by polsala**
## Rationale
- Automated proposal from the Gemini generator delivering a fresh community utility.

## Why safe to merge
- Utility is isolated to `utils/<util_name>`.
- README + tests ship together (see folder contents).
- No secrets or credentials touched.

## Test Plan
- Follow the instructions in the generated README

## Links
- Generated docs and examples committed alongside this change.

## Mock Justification
- Not applicable; generator did not introduce new mocks.

### CODE REVIEW

**Comment by polsala** [comment]
✅ What’s solid
*   **Robust File System Traversal:** The utility correctly employs `os.walk` with `topdown=False` for directory detection and sorts directories by length in `clean_up` before deletion. This ensures safe and accurate removal of nested empty directories, a critical detail for such utilities.
*   **Comprehensive Error Handling:** `try-except OSError` blocks are strategically placed in `find_empty_files` and `clean_up` to gracefully handle common file system issues like permission errors or broken symlinks. The `main` function also validates the input path, enhancing overall resilience.
*   **Clear User Experience:** The default dry-run mode, explicit `--delete` flag with confirmation, and verbose output options provide a safe and transparent user experience, allowing users to understand and control the utility's actions before making irreversible changes.
*   **Self-Contained and Portable:** The utility is a single Python script with standard library dependencies, making it highly portable and easy to run without complex installation requirements.

🧪 Tests
*   **Excellent Coverage:** The `test_duster.py` suite provides thorough coverage, testing core logic (`find_empty_files`, `find_empty_dirs`, `generate_report`), various deletion scenarios (dry run, successful delete, error handling during deletion), and the entire `main` function flow, including argument parsing and user interaction.
*   **Effective Mocking:** The tests make appropriate and extensive use of `unittest.mock` to isolate the code under test from actual file system operations and user input. This ensures tests are fast, reliable, and deterministic, especially crucial for a utility that performs destructive actions.
*   **Edge Case Consideration:** The test suite includes scenarios for broken symlinks, permission errors during deletion, and cases where no empty items are found, demonstrating a strong focus on robustness.

🔒 Security
*   **Principle of Least Privilege (Implicit):** The utility operates strictly within the specified `--path` and requires explicit user confirmation via the `--delete` flag before performing any destructive actions, significantly limiting its potential impact.
*   **Input Validation:** The `main` function validates the provided `--path` to ensure it points to an existing and valid directory, preventing accidental operations on incorrect or non-existent targets.
*   **Confirmation for Destructive Actions:** The interactive `input()` prompt before deletion is a vital security measure, preventing unintended data loss by requiring explicit user consent.

🧩 Docs/DX
*   **Clear README:** The `README.md` is well-structured and highly informative, offering a concise overview, detailed features, clear usage instructions with examples, and straightforward installation and testing guidance. This significantly enhances the developer and user experience.
*   **Helpful Command-Line Arguments:** The `argparse` setup provides intuitive argument names (`--path`, `--delete`, `--verbose`) and a clear `--help` message, making the utility easy to understand and use from the command line.
*   **Informative Reporting:** The `generate_report` function produces a user-friendly summary of detected empty files and directories, which is valuable for understanding the utility's findings in both dry-run and active modes.
*   **Overly Nested Directory Structure:** The current file path `utils/nightly-nightly-data-dust-bunny-dust/utils/nightly-data-dust-bunny-duster/` appears redundant. The `nightly-nightly-data-dust-bunny-dust` segment seems like an artifact of the generation process. Consider simplifying the path to `utils/nightly-data-dust-bunny-duster/` to improve clarity and reduce unnecessary nesting. This change would also need to be reflected in the `README.md`'s installation instructions.

    ```diff
    --- a/utils/nightly-nightly-data-dust-bunny-dust/utils/nightly-data-dust-bunny-duster/README.md
    +++ b/utils/nightly-data-dust-bunny-duster/README.md
    @@ -50,7 +50,7 @@
     1.  Navigate to the utility's directory:
         ```bash
         cd utils/nightly-data-dust-bunny-duster
    -    cd utils/nightly-nightly-data-dust-bunny-dust/utils/nightly-data-dust-bunny-duster
         ```
     2.  Run directly:
         ```bash
    ```

🧱 Mocks/Fakes
*   **Appropriate Mocking Strategy:** The tests effectively use `unittest.mock` to replace external dependencies such as file system operations (`os.walk`, `os.remove`, `os.rmdir`), user input (`builtins.input`), and argument parsing (`argparse.ArgumentParser.parse_args`). This ensures that tests are focused on the internal logic of the `duster.py` module.
*   **Clear Mock Rationale:** The comments within `test_duster.py` explicitly state the rationale for each mock, which significantly aids in understanding the test's intent and improves maintainability.
*   **Comprehensive Mocking for `main`:** The `test_main_*` functions demonstrate a well-orchestrated mocking strategy to simulate various end-to-end scenarios for the `main` function without interacting with the actual file system, which is crucial for testing a utility that performs destructive actions.
*   **Potential for `pathlib` Adoption:** While the `os` module is used effectively, consider exploring `pathlib` for future iterations. `pathlib` often provides a more object-oriented and readable way to handle file system paths, which can simplify code and reduce potential `os.path.join` errors, especially for complex path manipulations.

    ```python
    # Example: Using pathlib for path construction and checks
    from pathlib import Path

    def find_empty_files(root_dir: str) -> List[str]:
        root_path = Path(root_dir)
        empty_files = []
        for filepath in root_path.rglob('*'): # rglob for recursive search
            if filepath.is_file() and filepath.stat().st_size == 0:
                empty_files.append(str(filepath)) # Convert back to string if needed for os.remove
        return empty_files
    ```

**Review by polsalaTipoDeIncognito — APPROVED** [review]
✅ Auto-approved: All AI agent reviews passed and CI checks succeeded.
